Source/WebKit/WebProcess/GPU/graphics/RemoteImageBufferProxy.h

```diff
@@ -116,9 +114,13 @@ class RemoteImageBufferProxy : public WebCore::ConcreteImageBuffer<BackendType>
         m_receivedFlushIdentifierChangedCondition.notifyAll();
     }
 
-    void setNeedsFlush(bool needsFlush) final
+    void willUpdate() final
```

Comment:
This function will be called for all the recorded GraphicsContext functions including the state changing calls.  For these calls we do not need to invalidate the image cache.

- Do we need to call the fill-empty-rect-hack only for the drawing functions? Do need to call the hack from Recorder::appendStateChangeItemIfNecessary() instead of this function?
- Do we need to call the fill-empty-rect-hack always every time we flush the context? We know that a new NativeImage is created from the backend when copyNativeImage() is called. Can we use this information to call the fill-empty-rect-hack only after copyNativeImage() is called?
- Do we need to measure the perf impact of this patch?

Reply:
- Do we need to call ensureNativeImagesHaveCopiedBackingStore() only when canMapBackingStore() returns true.